File: scrt/script/wait2update.js
# $language = "javascript"
# $interface = "1.0"

var ret = null;
var prompt1 = "bootloader#";
var prompt2 = ">>";
var intr1 = "^C";
var intr2 = "<INTERRUPT>";
var progress1 = "Bytes";
var progress2 = "#";
var done1 = "[Done]";
var done2 = "OK";
var srvip = "192.168.100.106";//"10.1.1.2";
var myDate = new Date();
var cursec = myDate.getSeconds(); // [0,59]
var ip_pending = cursec + 190;
var localip = srvip.substring(0, srvip.lastIndexOf('.') + 1) + ip_pending; // [190,250]
while (ret == "" || ret == undefined || ret == null) {
	crt.Screen.Send(" \r\n");
	ret = crt.Screen.WaitForStrings(prompt1, prompt2, intr1, intr2, 1);
}

ret = ((ret == 1 || ret == 2) ? null : -1);
//crt.Screen.Send("#ret "+ret+"\r\n");
while (ret == "" || ret == undefined || ret == null) {
	crt.Screen.Send("update_rootfs " + srvip + " " + localip + " rootfs.ubi\r\n");
	ret = crt.Screen.WaitForStrings(progress1, progress2, intr1, intr2, 3);
}

ret = ((ret == 1 || ret == 2) ? null : -1);
while (ret == "" || ret == undefined || ret == null) {
	ret = crt.Screen.WaitForStrings(prompt1, prompt2, done1, done2, intr1, intr2, 300);
}

if (ret == 3 || ret == 4) {
	crt.Screen.Send("reset\r\n");
} else {
	crt.Screen.Send("#err " + ret + "\r\n");
}

File: scrt/script/wait2update.py
import sys, os, time, json

(strScriptPath, strScriptName) = os.path.split(__file__)
if strScriptPath not in sys.path:
    sys.path.insert(0, strScriptPath)
sys.dont_write_bytecode = True

from utils import session

sess = session.sess(crt.GetActiveTab())
srvip = "2.2.2.106" if crt.GetActiveTab().Caption.find("serial") >= 0 else "192.168.100.106"
localip = ""  # calc by srvip

opts = json.loads(crt.Arguments[0]) if crt.Arguments.Count == 1 else {}
srvip = opts.get("ip", srvip)


def genLip(crt):
    global localip
    cursec = (int(time.time())) % 60  # [0,59]
    ip_pending = str(cursec + 190)
    localip = srvip[0 : srvip.rindex(".") + 1] + ip_pending  # [190,250]


def exec_upgrade(crt):
    loop = True
    if not sess.is_uboot() and not sess.wait2uboot():
        crt.Screen.Send("\3\r\n# script terminated for u.\r\n")
        loop = False

    while loop:
        crt.Screen.Send("update_rootfs0 " + srvip + " " + localip + " rootfs.ubi\r\n")
        ret = sess.wait2exec(["####"], 20, "")
        if ret > 1:
            crt.Screen.Send("\3\r\n# script terminated for u.\r\n")
            break
        if ret == 0:
            crt.Screen.Send("\3\3\3# retry later\r\n")  # ctrl+c to break current update
            crt.Screen.WaitForStrings(["####"], 1)  # escape ctrl+c strings
            continue
        ret = sess.wait2exec(["written: OK"], 300, "reset\n")
        if ret > 1:
            crt.Screen.Send("\3\r\n# script terminated for u.\r\n")
        if ret == 0:
            crt.Screen.Send("# fatal error occurred\r\n")
        break


genLip(crt)
exec_upgrade(crt)
